### ISSUE TITLE
Source Generation Based Configuration

### ISSUE THREAD

**Issue by chyyran** [SnowflakePowered/snowflake#710]
## Preamble

This document outlines a rewrite of `Snowflake.Configuration`, keeping a compatible or similar API, that removes the need for dynamic proxies and simplifies the internal configuration plumbing, by using [C# Source Generators](https://devblogs.microsoft.com/dotnet/introducing-c-source-generators/) to create implementations of configuration interfaces at runtime.

This issue is dependent on resolution of #664.

## Goals

* Simplify and speed up instantiation and creation of Configuration instances
* Simplify the database schema for Configuration storage
* Remove Castle dependency

## Specification
We want to make minimal changes to the API of `IConfigurationCollection`, `IConfigurationSection`, and `IConfigurationOption`, but usage API changes are acceptable. That means that changes to `ConfigurationOptionAttribute`, `ConfigurationSectionAttribute`, etc. are acceptable, but the 'flat' API of `IConfigurationCollection` and `IConfigurationSection` should remain as is.

`IConfigurationValueCollection` should also remain fairly similar, but we should be open to any changes regarding this. 

### ConfigurationOption
Instead of a single configuration option attribute, we will now provide for attributes for different types of options.

#### Before
```csharp
interface MyConfiguration: IConfigurationSection<MyConfiguration> {
  [ConfigurationOption("someOption", false)]
  bool SomeOption { get; set; }
   
   [ConfigurationOption("someStringOption", "default value")]
   string SomeStringOption { get; set; }
}
```

#### After
```csharp
interface MyConfiguration: IConfigurationSection<MyConfiguration> {
  [BooleanOption("someOption", false)]
  bool SomeOption { get; set; }
   
   [StringOption("someStringOption", "default value")]
   string SomeStringOption { get; set; }
}
```

### ConfigurationSection
Currently, `ConfigurationSection` instances are populated at runtime by a target-less interface proxy generator. With source generators, we instead handle this at compile time.

#### Generation
Consider a configuration section defined like so

```csharp
[ConfigurationSection("myconfig", "myconfig")]
partial interface MyConfiguration {
  [BooleanOption("myOption", false)]
  bool MyBoolean { get; set; }
}
```

This will generate the following amendment to the interface

```csharp
// SnowflakeInternalSectionInstance type will change on rebuild, but that's fine.
[SnowflakeInternalSectionInstance(typeof(MyConfigurationProxy_a2hs4a))]
partial interface MyConfiguration { 
  ISectionDescriptor Descriptor { get; }
}

// This is the proxy class that is generated
public class MyConfigurationProxy_a2hs4a : MyConfiguration {
  public MyConfigurationProxy_a2hs4a(ISectionDescriptor sectionDescriptor, IConfigurationValueCollection collection) 
  {
    this.__sectionDescriptor = sectionDescriptor;
    this.__backingCollection = collection;
    this.PopulateDefaults();
  }

   private void PopulateDefaults() {
   // populate backingcollection[this.__sectionDescriptor] here.
   }
   ISectionDescriptor Descriptor => this.__sectionDescriptor { get; }
  bool MyConfiguration.MyBoolean { 
    get { return this.__backingCollection[this.__sectionDescriptor, nameof(MyConfiguration.MyBoolean)]; }
    set { this.__backingCollection[this.__sectionDescriptor, nameof(MyConfiguration.MyBoolean)] = value; }
  }
}
```

#### Instantiation
Instantiating a section will involve the following factory
```csharp
class ConfigurationSection<T> {
  T Section { get; private set; }
  ISectionDescriptor Descriptor { get; }
  private T MakeSection(IConfigurationValueCollection collection) {
    return (T)Instantiate.MakeInstance(
    typeof(T).GetAttribute<SnowflakeInternalSectionInstance>().Type,
    new [] { typeof(ISectionDescriptor), typeof(IConfigurationValueCollection) },
    Expression.Constant(ISectionDescriptor), Expression.Constant(collection);
  }
}
```

## ConfigurationCollection
Similarly with configuration sections, consider this example configuration
```csharp
partial interface MyConfigurationCollection {
  [ConfigurationTargetMember("#mytarget")]
  MyConfiguration MyConfiguration { get; set; }
}
```
#### Generation
We generate the following amendment
```csharp
[SnowflakeInternalConfigurationInstance(typeof(MyConfigurationCollectionProxy_b34dfa))]
partial interface MyConfiguration { }
```

and the following proxy

```csharp
class MyConfigurationCollectionProxy_b34dfa: MyConfigurationCollection, IEnumerable<IConfigurationSectionDescriptor> {
    public MyConfigurationCollectionProxy_b34dfa(IConfigurationValueCollection collection) {
      this.__backingCollection = collection;
      this.MyConfiguration = new ConfigurationSection<MyConfiguration>(collection);
      this.__descriptors = new List[] {
        this.MyConfiguration.Descriptor
       } // this should be immutable list in actuality
   }
   private ConfigurationSection<MyConfiguration> MyConfiguration { get; }
   MyConfiguration MyConfigurationCollection.MyConfiguration => MyConfiguration.Section;
   IEnumerator<IConfigurationSectionDescriptor>.GetEnumerator() => this.__descriptors.GetEnumerator();
}
```

#### Instantiation
Similarly...

```csharp
class ConfigurationCollection<T>: IEnumerable<IConfigurationSectionDescriptor> {
  T Configuration { get; private set; }
  private T MakeCollection(IConfigurationValueCollection collection) {
    return (T)Instantiate.MakeInstance(
    typeof(T).GetAttribute<SnowflakeInternalCollectionInstance>().Type,
    new [] { typeof(IConfigurationValueCollection) },
     Expression.Constant(collection);
  }
  public IEnumerator<IConfigurationSectionDescriptor> GetEnumerator() => ( (IEnumerable<IConfigurationSectionDescriptor>)this.Configuration).GetEnumerator();
}
```

**Comment by chyyran** [comment]
work  underway in feature branch https://github.com/chyyran/snowflake/tree/feature-srcgen-config

**Comment by chyyran** [comment]
Fixed in #819
